sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprVarNode.java

```diff
@@ -0,0 +1,137 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.runtime.debug.debugexpr.nodes;
+
+import java.util.List;
+
+import com.oracle.truffle.api.Scope;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.InvalidArrayIndexException;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.llvm.runtime.LLVMLanguage;
+import com.oracle.truffle.llvm.runtime.debug.LLVMDebuggerValue;
+import com.oracle.truffle.llvm.runtime.debug.debugexpr.parser.DebugExprException;
+import com.oracle.truffle.llvm.runtime.debug.debugexpr.parser.DebugExprType;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+
+public class DebugExprVarNode extends LLVMExpressionNode {
+
+    private final String name;
+    private Iterable<Scope> scopes;
+    private DebugExprType type;
+    private Object value;
+    private Object member;
+    private Object metaObj;
+
+    public DebugExprVarNode(String name, Iterable<Scope> scopes) {
+        this.name = name;
+        this.scopes = scopes;
+        this.type = null;
+        this.value = null;
+        this.member = null;
+        this.metaObj = null;
+        findMemberAndType();
+    }
+
+    private void findMemberAndType() {
+        LLVMLanguage.getLLVMContextReference().get();
+        InteropLibrary library = InteropLibrary.getFactory().getUncached();
+        for (Scope scope : scopes) {
+            Object vars = scope.getVariables();
+            library.hasMembers(vars);
+            try {
+                final Object memberKeys = library.getMembers(vars);
+
+                if (library.isMemberReadable(vars, name)) {
+                    member = library.readMember(vars, name);
+                    LLVMDebuggerValue ldv = (LLVMDebuggerValue) member;
+                    metaObj = ldv.getMetaObject();
+                    type = DebugExprType.getTypeFromSymbolTableMetaObject(metaObj);
+                    return;
+                }
+                if (library.hasArrayElements(memberKeys) && library.isMemberReadable(memberKeys, name)) {
```

Comment:
That looks wrong. `memberKeys` should always be an array (but no harm in checking), but then `isMemberReadable` will probably always return `false`, since `memberKeys` is an array, not a struct. Are you sure this `if` is ever entered?